### DISCUSSION TITLE
Filament 3.2, Repeater value Not shown in edit mode.

### DISCUSSION THREAD

**Discussion by mohsinkhanwork** [filamentphp/filament#11080] [Help]
### Package

Form builder

### Package Version

v3.x

### How can we help you?

I have dependent fields in the question model. when I select type the question form gets changed based on type ( int values 1,2..). The issue is it works best on the create page but when it goes to the edit page, the repeater's option values do not show. Only the select field is selected to the single choice question but values from the repeater do not gets shown, My code is below:

![Screenshot 2024-01-26 164254](https://github.com/filamentphp/filament/assets/84306156/9bba753b-0303-42f2-a19c-b29850bb4f23)

**Comment by leandrocfe** [comment]
Are you using a json field to store the repeater data? Share your model please (code not screenshot)

**Reply by mohsinkhanwork** [reply]
yes, I am using the details column as JSON where I'm saving the question detail and options details, they are saving fine as expected but after I save. The page goes to the edit page and then the selected option is selected like a single-choice question in Select but the repeater options are not populated. below I'm sharing the model, db saved data, and the code. for testing, I used only scq option. the afterStateHydrated works fine only when I change the select type from other to scq but when I refresh the page it does not populate the options in the repeater.

**this is the question model:**
`
class Question extends TenantModel
{
    protected $fillable = [
        'text',
        'type',
        'points',
        'difficulty',
        'details',
        'subject_id',
    ];

    /**
     * The attributes that should be cast.
     *
     * @var array<string, string>
     */
    protected $casts = [
        'type' => QuestionType::class,
        'difficulty' => QuestionDifficulty::class,
        'details' => 'json',
    ];

    //------------------------------------------------------
    // Accessors
    //------------------------------------------------------
    // Here you can define attribute accessors
    // e.g., formatting the text or details

    //------------------------------------------------------
    // Mutators
    //------------------------------------------------------
    // Here you can define attribute mutators
    // e.g., custom logic before saving data

    //------------------------------------------------------
    // Relationships
    //------------------------------------------------------

    /**
     * Question belongs to a Subject.
     */
    public function subject(): BelongsTo
    {
        return $this->belongsTo(Subject::class);
    }

    /**
     * Question belongs to many Topics.
     */
    public function topics(): BelongsToMany
    {
        return $this->belongsToMany(Topic::class, 'question_topics')->withTimestamps();
    }
}`

**this is my code:**
`public static function form(Form $form): Form
    {
        return $form
            ->schema([
                Group::make()
                ->schema([
                  
                    Forms\Components\Section::make(__('entity.question.Question_options'))
                        ->schema([
                            Select::make('type')
                                    ->options(QuestionType::getFilterableOptions())
                                    ->label(__('entity.question.type'))
                                    ->live()
                                    ->afterStateUpdated(fn (Select $component) => $component
                                            ->getContainer()
                                            ->getComponent('questionTypeFields')
                                            ->getChildComponentContainer()
                                            ->fill()),
                            Grid::make(1)
                            ->schema(fn (Get $get): array => match ($get('type')) {
                                    '1' => [                // scq
                                        Repeater::make('details.options')
                                            ->label(__('entity.question.options'))
                                            ->schema([
                                                Hidden::make('optionID')
                                                        ->default(function () {
                                                            return 'ulid-' . Str::random(10 );
                                                        }),
                                                TextInput::make('text')
                                                    ->label(__('entity.question.options'))
                                                    ->columnSpan(1),
                                                Radio::make('isCorrect')
                                                    ->options([
                                                        'true' => __('entity.question.correct'),
                                                        'false' => __('entity.question.incorrect'),
                                                    ])
                                                    ->inline()
                                                    ->required(),
                                                FileUpload::make('attachments')
                                                    ->label(__('entity.question.image'))
                                                    ->disk('s3')
                                                    ->directory('options_attachments')
                                                    ->visibility('private')
                                                    ->columnSpan(1),
                                            ])
                                            ->live()
                                            ->afterStateHydrated(function ($context, $state, $component) {
                                                Log::info('test 1 ' . $context, $state, $component);
                                                if ($context === 'edit' && !empty($state)) {
                                                    Log::info('test 2 ' . $context, $state, $component);
                                                    $options = $component->getRecord()->details['options'];
                                                    foreach ($options as &$option) {
                                                        $option['attachments'] = explode(',', $option['attachments']); // Assuming a comma-separated list
                                                        // Log::info('C' .$option['attachments']);
                                                    }
                                                    
                                                    $component->state($options);
                                                }
                                                Log::info('D ' .$component->getRecord());
                                            })
                                    ],
                                    default => [],
                            })
                                    ->key('questionTypeFields'),
                            ])
                            ->columnSpan(1),
                        ])
                ->columnSpan(['lg' => 2]),
        ])
        ->columns(3); `
        
        **and this is the data that gets saved.:**
        
    `    `{
  "attachments": "question_attachments\/01HN0Y8ZRVECWXN1WCQ0WQAN5Z.png",
  "options": [{
    "optionID": "ulid-4XcUBc1RQu",
    "text": "1122",
    "isCorrect": "true",
    "attachments": "options_attachments\/01HN12RGQ6H71YQDMVN6Z79G5J.png"
  }, {
    "optionID": "ulid-9Yk6Bvp6bv",
    "text": "332211",
    "isCorrect": "false",
    "attachments": "options_attachments\/01HN12RGR1EK3G5TNGH5VCAFXS.jpg"
  }]
}``

**Reply by mohsinkhanwork** [reply]
@leandrocfe brother any help in such case? i shared my code above of db saved data, resource and the model

**Reply by leandrocfe** [reply]
Try `$casts = ['details' => 'array']` instead of `json`

**Reply by mohsinkhanwork** [reply]
hello, I changed the details to the array but the repeater options are not populating when I refresh the edit page. please check all my code above which I'm doing to get the options of a repeater on the edit page based on select.

**Comment by mohsinkhanwork** [comment]
hi so the issue was in my code, in the edit, the page type was taking 1 but when I changed the select type from the select drop-down then it was taking 1 as a string like "1" which is why options were not appearing. I solved it by removing strings like "1" -> 1 and added int strict type. like below

`Grid::make(1)
                    ->schema(fn (Get $get): array => match ((int) $get('type')) {
                        1 => [                // scq
                            Repeater::make('details.options')
                                ->la`
